command/timeout.go

```diff
@@ -22,20 +34,34 @@ func (c *timeout) Name() string     { return "timeout.update" }
 
 // ParseParams parses the params into the the timeout struct.
 func (c *timeout) ParseParams(params map[string]interface{}) error {
-	err := mapstructure.Decode(params, c)
-	if err != nil {
-		return errors.Wrap(err, "error decoding params")
-	}
-	if c.TimeoutSecs == 0 && c.ExecTimeoutSecs == 0 {
-		return errors.New("Must set at least one timeout")
-	}
-
+	c.params = params
 	return nil
 }
 
 // Execute updates the idle timeout.
 func (c *timeout) Execute(ctx context.Context,
 	comm client.Communicator, logger client.LoggerProducer, conf *model.TaskConfig) error {
+	// do the params parsing here rather than in ParseParams because we want
+	// to destructure only if parsing as ints fails.
+	if err := mapstructure.Decode(c.params, c); err != nil ||
+		(c.TimeoutSecs == 0 && c.ExecTimeoutSecs == 0) {
+		// If destructuring as ints fails, or neither timeout is set, destructure as strings.
+		t := &timeoutStr{}
+		if errStr := mapstructure.Decode(c.params, t); errStr != nil {
+			return errors.New("could not decode params as either string or int")
+		}
+		if err := util.ExpandValues(t, conf.Expansions); err != nil {
+			return errors.Wrap(err, "error expanding expansion values")
+		}
+		timeout, errTimeout := strconv.Atoi(t.TimeoutSecs)
+		exec, errExec := strconv.Atoi(t.ExecTimeoutSecs)
+		if errTimeout != nil && errExec != nil {
+			return errors.Errorf("could not convert strings to int, %s, %s", t.TimeoutSecs, t.ExecTimeoutSecs)
+		}
+		c.TimeoutSecs = timeout
+		c.ExecTimeoutSecs = exec
+	}
```

Comment:
it might make it easier to have this if-block in a private method and save private int value versions of the Atoi'd versions of these values, so that you can test them separately.

Reply:
I think in general the two test cases (an expansion, which is a string, and an int, not expanded) cover this.